### PR TITLE
Add "Mark as Complete" Feature with Local Storage Progress

### PR DESCRIPTION
**Pull request by jrakibi**
Adds ability for users to mark topics as completed and see their learning progress. All data is stored in the user's browser (localStorage) with no external tracking.

Key changes:
- "Mark as Complete" button on topics
- Visual completion indicators in navigation
- Progress visualization in hero section
- Next topic suggestions after completion

image

image

### CODE REVIEW

**Comment by kouloumos** [comment]
Does this checks if the user have actually completed the exercises before allowing them to mark it as complete?

Idea: in non-exercise pages, maybe a question could come up before allowing them to mark it as complete.

Note: It seems that you need to do some kind of rebase here

**Comment by jrakibi** [comment]
- Rebased, thanks 
- Users can manually mark any topic as complete if they choose
- Implementing a check for exercise solutions, especially for Python exercises hosted on third-party platforms like Trinket, would be challenging since we don't have direct access to the state
- For now, we can keep the feature as-is to ensure we stay on schedule with the upcoming Bitcoin Indonesia cohort

cc @kouloumos

File: components/bitcoin-topics/layouts/TopicBanner.tsx [hunk 1]
```diff
@@ -82,6 +197,23 @@ export default function TopicBanner({
         return () => window.removeEventListener("scroll", handleScroll)
     }, [])
 
+    useEffect(() => {
+        const handleCompletionChange = () => {
+            const saved = localStorage.getItem("completedTopics")
+            if (saved) {
+                setCompletedTopics(JSON.parse(saved))
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/layouts/TopicBanner.tsx:204]: There is no `completedTopics` State or setter `setCompletedTopics`

**Review by Emmanuel-Develops — CHANGES_REQUESTED** [review]

File: components/bitcoin-topics/layouts/TopicBanner.tsx (hunk 1 above)
Comment (jrakibi) [reply components/bitcoin-topics/layouts/TopicBanner.tsx]: fixed

File: components/bitcoin-topics/layouts/TopicBanner.tsx [hunk 2]
```diff
@@ -245,9 +379,8 @@ export default function TopicBanner({
                                 summary={content.summary}
                             />
                             <Prose>{children}</Prose>
-                            {/* Add the GitHub edit button for mobile */}
                             <div className="max-w-3xl mx-auto mt-8 flex justify-end">
-                                <EditOnGitHubButton />
+                                <ActionButtons />
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/layouts/TopicBanner.tsx:383]: Remember the discussion on `TopicBanner` and `TopicLayout`.

Topics using `TopicLayout` won't have the <ActionButtons /> component, ergo won't have the `Mark as Complete` button.

It seems only Bitcoin History will be affected by this change as it uses `TopicLayout`.

File: components/bitcoin-topics/Navigation.tsx [hunk 3]
```diff
@@ -11,15 +11,24 @@ import {
...
+import {
+    FaChevronRight,
+    FaChevronDown,
+    FaCheck,
+    FaUsers,
+    FaCalendar
+} from "react-icons/fa"
 import * as Icons from "react-icons/fa"
+import { HiArrowRight } from "react-icons/hi2"
+import Image from "next/image"
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/Navigation.tsx:23]: Never used

File: components/bitcoin-topics/Navigation.tsx [hunk 4]
```diff
@@ -11,15 +11,24 @@ import {
 import { Topic, allTopics } from "@/.contentlayer/generated"
 import { useState, useEffect, useCallback, useMemo } from "react"
 import { motion } from "framer-motion"
-import { FaChevronRight, FaChevronDown } from "react-icons/fa"
+import {
+    FaChevronRight,
+    FaChevronDown,
+    FaCheck,
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/Navigation.tsx:17]: Never used

File: components/bitcoin-topics/Navigation.tsx [hunk 5]
```diff
@@ -94,6 +106,57 @@ export function Navigation({
...
+                setCompletedTopics(completedSet)
+            } catch (error) {
+                console.error("Error parsing completedTopics:", error)
+            }
+        }
+    }, [])
+
+    // Add listener for completion changes
+    useEffect(() => {
+        const handleCompletionChange = () => {
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/Navigation.tsx:128]: this function is duplicated in `ln:111`, we can have the function outside the useEffect so the useEffects can consume it.

Also, I don't see the need of the first useEffect. This useEffect to handle `topicCompletionChanged` event is going to run on initial render, so the first one is unneeded.
Was there a specific need for the first useEffect?

File: components/bitcoin-topics/Navigation.tsx [hunk 6]
```diff
@@ -11,15 +11,24 @@ import {
 import { Topic, allTopics } from "@/.contentlayer/generated"
 import { useState, useEffect, useCallback, useMemo } from "react"
 import { motion } from "framer-motion"
-import { FaChevronRight, FaChevronDown } from "react-icons/fa"
+import {
+    FaChevronRight,
+    FaChevronDown,
+    FaCheck,
+    FaUsers,
+    FaCalendar
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/Navigation.tsx:19]: Never used

File: components/bitcoin-topics/Navigation.tsx [hunk 7]
```diff
@@ -11,15 +11,24 @@ import {
...
+} from "react-icons/fa"
 import * as Icons from "react-icons/fa"
+import { HiArrowRight } from "react-icons/hi2"
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/Navigation.tsx:22]: Never used

File: components/bitcoin-topics/layouts/TopicBanner.tsx [hunk 8]
```diff
@@ -31,28 +31,143 @@ export default function TopicBanner({
     prev,
     children
 }: LayoutProps) {
-    const pathname: string | null = usePathname()
-    const isHomePage: boolean = pathname === "/decoding"
+    const router = useRouter()
```
Comment (Emmanuel-Develops) [inline components/bitcoin-topics/layouts/TopicBanner.tsx:34]: router is not used

File: components/bitcoin-topics/layouts/TopicBanner.tsx (hunk 2 above)
Comment (jrakibi) [reply components/bitcoin-topics/layouts/TopicBanner.tsx]: yep, refactored now

File: components/bitcoin-topics/Navigation.tsx (hunk 5 above)
Comment (jrakibi) [reply components/bitcoin-topics/Navigation.tsx]: correct 
I combined these two effects since they both deal with the `handleCompletionChange` function

**Comment by jrakibi** [comment]
@Emmanuel-Develops I believe all feedback has been addressed

File: components/bitcoin-topics/Navigation.tsx (hunk 4 above)
Comment (Emmanuel-Develops) [reply components/bitcoin-topics/Navigation.tsx]: hey @jrakibi this is still unused.
We can tighten up `eslint` so it catches stuff like this (in another PR)

**Comment by Emmanuel-Develops** [comment]
> @Emmanuel-Develops I believe all feedback has been addressed

Yup yup looks good!

**Review by Emmanuel-Develops — APPROVED** [review]

**Pull request merged**